test_util/dcos_api_session.py

```diff
@@ -89,31 +68,92 @@ def api_request(self, method, path_extension, *, scheme=None, host=None, query=N
-        """Proxy class for DC/OS clusters.
+    def __init__(
+            self,
+            dcos_url: str,
+            masters: Optional[List[str]],
+            slaves: Optional[List[str]],
+            public_slaves: Optional[List[str]],
+            default_os_user: str,
+            auth_user: Optional[DcosUser]):
+        """Proxy class for DC/OS clusters. If any of the host lists (masters,
+        slaves, public_slaves) are provided, the wait_for_dcos function of this
+        class will wait until provisioning is complete. If these lists are not
+        provided, then there is no ground truth and the cluster will be assumed
+        the be in a completed state.
 
         Args:
             dcos_url: address for the DC/OS web UI.
             masters: list of Mesos master advertised IP addresses.
-            public_masters: list of Mesos master IP addresses routable from
-                the local host.
             slaves: list of Mesos slave/agent advertised IP addresses.
+            public_slaves: list of public Mesos slave/agent advertised IP addresses.
             default_os_user: default user that marathon/metronome will launch tasks under
             auth_user: use this user's auth for all requests
                 Note: user must be authenticated explicitly or call self.wait_for_dcos()
         """
         super().__init__(Url.from_string(dcos_url))
-        self.masters = sorted(masters)
-        self.public_masters = sorted(public_masters)
-        self.slaves = sorted(slaves)
-        self.public_slaves = sorted(public_slaves)
-        self.all_slaves = sorted(slaves + public_slaves)
+        self.master_list = masters
+        self.slave_list = slaves
+        self.public_slave_list = public_slaves
         self.default_os_user = default_os_user
         self.auth_user = auth_user
 
-        assert len(self.masters) == len(self.public_masters)
+    @staticmethod
+    def get_args_from_env():
+        """ Provides the required arguments for a unauthenticated cluster
+        """
+        masters = os.getenv('MASTER_HOSTS')
+        slaves = os.getenv('SLAVE_HOSTS')
+        public_slaves = os.getenv('PUBLIC_SLAVE_HOSTS')
+        return {
+            'dcos_url': os.getenv('DCOS_DNS_ADDRESS', 'http://leader.mesos'),
+            'masters': masters.split(',') if masters else None,
+            'slaves': slaves.split(',') if slaves else None,
+            'public_slaves': public_slaves.split(',') if public_slaves else None,
+            'default_os_user': os.getenv('DCOS_DEFAULT_OS_USER', 'root')}
+
+    @property
+    def masters(self):
+        return sorted(self.master_list)
+
+    @property
+    def slaves(self):
+        return sorted(self.slave_list)
+
+    @property
+    def public_slaves(self):
+        return sorted(self.public_slave_list)
+
+    @property
+    def all_slaves(self):
+        return sorted(self.slaves + self.public_slaves)
+
+    def set_node_lists_if_unset(self):
```

Comment:
The expected behavior is to fill in whatever is unset, i.e. the former option. The intention is that perhaps you only care about making sure the public agnents provisioned, then you can provide just that value and the other nodes will be taken from DC/OS so you can run your test. I think the logging statements already communicate enough but I do not mind bumping them up a level